internal/commands/relationship.go

```diff
@@ -250,100 +279,197 @@ func readRelationships(cmd *cobra.Command, args []string) error {
+// and returns the fields representing the 3 arguments. This is to handle
+// the fact that relationships specified via stdin can't escape spaces like
+// shell arguments.
+func parseRelationshipLine(line string) ([]string, error) {
+	resourceIdx := strings.IndexFunc(line, unicode.IsSpace)
+	if resourceIdx == -1 {
+		return nil, fmt.Errorf("expected %s to have 3 arguments, but got 0", line)
+	}
+
+	resource := line[:resourceIdx]
+	rest := strings.TrimSpace(line[resourceIdx+1:])
+	relationIdx := strings.IndexFunc(rest, unicode.IsSpace)
+	if relationIdx == -1 {
+		return nil, fmt.Errorf("expected %s to have 3 arguments, but got 1", line)
+	}
+
+	relation := rest[:relationIdx]
+	rest = strings.TrimSpace(rest[relationIdx+1:])
+	if rest == "" {
+		return nil, fmt.Errorf("expected %s to have 3 arguments, but got 2", line)
+	}
+
+	return []string{resource, relation, rest}, nil
+}
+
+type nextRelationshipFunc func() ([]string, error)
+
+func streamArgsFromFile(file *os.File) nextRelationshipFunc {
+	scanner := bufio.NewScanner(file)
+	return func() ([]string, error) {
+		if scanner.Scan() {
+			args, err := parseRelationshipLine(scanner.Text())
+			if err != nil {
+				return nil, err
+			}
+			return args, nil
+		}
+		return nil, scanner.Err()
+	}
+}
+
+func streamArgsFromCliArgs(args []string) nextRelationshipFunc {
+	ran := false
+	return func() ([]string, error) {
+		if ran {
+			return nil, nil
+		}
+		ran = true
+		return args, nil
+	}
+}
+
+func writeUpdates(ctx context.Context, grpcClient client.Client, updates []*v1.RelationshipUpdate, json bool) error {
+	if len(updates) == 0 {
+		return nil
+	}
+	request := &v1.WriteRelationshipsRequest{
+		Updates:               updates,
+		OptionalPreconditions: nil,
+	}
+
+	log.Trace().Interface("request", request).Msg("writing relationships")
+	resp, err := grpcClient.WriteRelationships(ctx, request)
+	if err != nil {
+		return err
+	}
+
+	if json {
+		prettyProto, err := PrettyProto(resp)
 		if err != nil {
 			return err
 		}
 
-		relation := args[1]
+		console.Println(string(prettyProto))
+	} else {
+		console.Println(resp.WrittenAt.GetToken())
+	}
+
+	return nil
+}
 
-		subjectNS, subjectID, subjectRel, err := ParseSubject(args[2])
+func toRelationship(f nextRelationshipFunc) (*v1.Relationship, error) {
+	args, err := f()
+	if err != nil {
+		return nil, err
+	}
+	if args == nil {
+		return nil, nil
+	}
+	return argsToRelationship(args)
+}
+
+func writeRelationshipCmdFunc(operation v1.RelationshipUpdate_Operation, clientFunc func(command *cobra.Command) (client.Client, error), input *os.File) func(cmd *cobra.Command, args []string) error {
+	return func(cmd *cobra.Command, args []string) error {
+		// getNextRelationship is a function that will fetch the next relationship to write.
+		// Either it will read a stream of relationships from stdin, or it will simply return
+		// the command line arguments once. When there are no more relationships, it will return
+		// nil for the first parameter.
```

Comment:
Why not make this a bit more explicit in the type system and create a place where this comment can live:

```go
// RelationshipParser is a closure that can produce relationships.
// When there are no more relationships, it will return ErrExhaustedRelationships.
type RelationshipParser func() (*v1.Relationship, error)

// ErrExhaustedRelationships signals that the last producible value of a RelationshipParser
// has already been consumed.
// Functions should return this error to signal a graceful end of input.
var ErrExhaustedRelationships = errors.New("exhausted all relationships")
```

```go
func FileRelationshipParser(f *os.File) RelationshipParser {
	scanner := bufio.NewScanner(file)
	return func() (*v1.Relationship, error) {
		if scanner.Scan() {
			res, rel, subj, err := parseRelationshipLine(scanner.Text()) // Notice this signature change
			if err != nil {
				return nil, err
			}
			return tupleToRel(res, rel, subj) // This does your sprintf + string parsing
		}
		return nil, scanner.Err()
	}
}
```

If you refactor the args func like this too, the call site gets cleaned up a bit:

```go
	return func(cmd *cobra.Command, args []string) error {
		parse := SliceRelationshipParser(args)
		if !isFileTerminal(input) {
			// If it's not a terminal, assume the program is being piped into.
			parse = FileRelationshipParser(input)
		}
		...
		for {
			rel, err := parse()
			if errors.Is(err, ErrRelationshipsExhausted) {
				break
			} else if err != nil {
				return err
			}
			...
		}

```

Reply:
I was a bit lazy to refactor this a bit further, thanks for doing it!